Feature: Add invoke lambda function operator

schirag1993 opened the pull request:
related: [#21009](https://github.com/apache/airflow/issues/21009)

This change is to implement AWS Lambda function invoke operator for the AWS provider package.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

schirag1993 committed:
add invoke lambda function operator

malthe commented on airflow/providers/amazon/aws/hooks/lambda_function.py:
```diff
@@ -40,30 +40,31 @@ class LambdaHook(AwsBaseHook):
+            Role=role,
+            Handler=handler,
+            Code=code,
+            **{k: v for k, v in kwargs.items() if v is not None},
```
What are these `kwargs` – the method is easier to use if they're spelled out to the user.

schirag1993 replied:
I understand. The changes have been made @malthe.

malthe commented on airflow/providers/amazon/aws/hooks/lambda_function.py:
```diff
@@ -40,30 +40,31 @@ class LambdaHook(AwsBaseHook):
         super().__init__(*args, **kwargs)
 
-    def invoke_lambda(self, payload: str) -> str:
+    def invoke_lambda(self, function_name: str, **kwargs):
```
It would be better if the parameters were spelled out, especially because they now also have typing information. But previously, those parameters also had defaults.

schirag1993 replied:
@malthe, I agree. I've made the recommended changes and added defaults.

malthe commented on airflow/providers/amazon/aws/operators/aws_lambda.py:
```diff
@@ -0,0 +1,121 @@
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.FunctionName = function_name
```
This should be `function_name`, etc.

schirag1993 replied:
@malthe, done!

malthe commented on airflow/providers/amazon/aws/operators/aws_lambda.py:
```diff
@@ -0,0 +1,121 @@
+            response = hook.invoke_lambda(
+                function_name=self.FunctionName,
+                **{
+                    key: self.__dict__[key]
```
This should be refactored a bit and the `kwargs` extraction can be put outside of the `try` handler.

schirag1993 replied:
@malthe, I understand. This has been refactored. Kindly review and resolve the conversation if the changes have addressed the comment.

malthe commented on airflow/providers/amazon/aws/operators/aws_lambda.py:
```diff
@@ -0,0 +1,121 @@
+                    )
+                },
+            )
+            self.log.info("Lambda response metadata: %s", json.dumps(response.get("ResponseMetadata")))
```
I don't think you need to serialize this as JSON; a Python repr string would be fine here:
```python
self.log.info("Lambda response metadata: %r", response.get("ResponseMetadata"))
```
Also, there's a lot of repetition in the error handling here.

schirag1993 replied:
@malthe, thank you for the review. I've made the changes for error handling. 

However, I am unable to think of a way to reduce the error handling bit. The errors are handled based on the response from the AWS API. If an exception is not thrown by the underlying boto3 SDK, the handling of the same becomes conditional leading to the multiple `if` blocks.

Would you have any recommendations?

edwardwang888 replied:
@schirag1993 I think @malthe may have been referring to the same thing as https://github.com/apache/airflow/pull/21686#discussion_r810686375, which is already resolved?

schirag1993 replied:
Yes, looks like it @edwardwang888. That should mark all comments addressed then.

malthe reviewed (commented):
I have raised a couple of code-level issues for your consideration.

schirag1993 committed:
AWS Lambda Invoke function operator

update operator and underlying hook code to explicitly specify invoke
and create function arguments

edwardwang888 commented on airflow/providers/amazon/aws/operators/aws_lambda.py:
```diff
@@ -0,0 +1,103 @@
+            self.log.info("Lambda response metadata: %r", response.get("ResponseMetadata"))
+        except Exception as e:
+            self.log.error(e)
+            raise e
```
This `except` block doesn't seem needed if you are simply re-printing the error?

schirag1993 replied:
You're right @edwardwang888, will make this change.

uranusjr commented on airflow/providers/amazon/aws/hooks/lambda_function.py:
```diff
@@ -40,30 +41,84 @@ class LambdaHook(AwsBaseHook):
+        client_context: Optional[str] = None,
+        payload: Optional[str] = None,
+        qualifier: Optional[str] = None,
+    ):
```
```suggestion
    def invoke_lambda(
        self,
        *,
        function_name: str,
        invocation_type: Optional[str] = None,
        log_type: Optional[str] = None,
        client_context: Optional[str] = None,
        payload: Optional[str] = None,
        qualifier: Optional[str] = None,
    ):
```

How about make these all keyword-only arguments? This is probably how a user is supposed to use them anyway (the invocation line would not be readable otherwise), so let’s enforce them.

Same for other functions on the hook.

schirag1993 replied:
Hello @uranusjr, these were originally kwargs. However, based on @malthe's feedback (https://github.com/apache/airflow/pull/21686#discussion_r810589155) from an earlier discussion (which I agree with), it may be difficult for first timers to understand what is required and the typing info helps as well. 
As we are building this for Airflow, it makes sense to spell out these parameters for users for better reference. 

What are your thoughts? If the above does not make sense, I'll go ahead and make the required changes.

uranusjr replied:
I do not suggest changing this to kwargs, but _keyword-only arguments_.

https://www.python.org/dev/peps/pep-3102/

malthe replied:
@schirag1993 yeah it's just adding a magic "*" argument to denote "you shall not pass" positional parameters after this.

schirag1993 replied:
Ah I misunderstood then. Thank you for the clarifications @uranusjr and @malthe! 
Will incorporate the changes.

schirag1993 replied:
Done @uranusjr! Thank you for the feedback. Today I learnt something new

edwardwang888 commented on tests/providers/amazon/aws/operators/test_lambda.py:
```diff
@@ -0,0 +1,107 @@
+        zip_output = io.BytesIO()
+        zip_file = zipfile.ZipFile(zip_output, "w", zipfile.ZIP_DEFLATED)
+        zip_file.writestr("lambda_function.py", code)
+        zip_file.close()
```
`ZipFile` can be used as a context manager:
```
with zipfile.ZipFile(...) as zip_file:
    zip_file.write_str(...)
```

schirag1993 replied:
+1, I'll make the required change @edwardwang888

schirag1993 replied:
Done

edwardwang888 commented on tests/providers/amazon/aws/operators/test_lambda.py:
```diff
@@ -0,0 +1,107 @@
+        )
+        return resp["Role"]["Arn"]
+
+    def create_lambda_function(self, function_name: str):
```
Technically, the helper functions (`create_zip`, `create_iam_role`, `create_lambda_function`) can be made static methods if they are not instance-specific?

schirag1993 replied:
I agree @edwardwang888. Will make this change.

schirag1993 replied:
Done!

edwardwang888 reviewed (commented):
Added some minor comments.

schirag1993 committed:
AWS Lambda Invoke Function Operator

modify test cases to use static method for lambda invoke operator
remove redundant try-except block

schirag1993 committed:
Update AWS Lambda Hook

update AWS lambda hook functions to enforce keyword only args

uranusjr reviewed (approved):
Note: This needs a major version bump in the Amazon provider.

schirag1993 committed:
add lambda function invoke operator

ferruzzi commented on docs/apache-airflow-providers-amazon/operators/lambda.rst:
```diff
@@ -0,0 +1,54 @@
+.. exampleinclude:: /../../airflow/providers/amazon/aws/example_dags/example_lambda.py
+    :language: python
+    :start-after: [START howto_lambda_operator]
+    :end-before: [END howto_lambda_operator]
```
```suggestion
.. exampleinclude:: /../../airflow/providers/amazon/aws/example_dags/example_lambda.py
    :language: python
    :dedent: 4
    :start-after: [START howto_lambda_operator]
    :end-before: [END howto_lambda_operator]
```

Nitpick since you indented the START/END tags.

ferruzzi commented on airflow/providers/amazon/aws/operators/aws_lambda.py:
```diff
@@ -0,0 +1,99 @@
+        self.log.info('Lambda function invocation succeeded: %r', response.get("ResponseMetadata"))
+        payload_stream = response.get("Payload")
+        payload = payload_stream.read().decode()
+        return payload
```
Seems like Lines 97/98 can be moved before the if "FunctionError" block and reused inside the block as well, no? You are getting and decoding the stream whether there is an error or not.

```suggestion
        payload_stream = response.get("Payload")
        payload = payload_stream.read().decode()
        if "FunctionError" in response:
            raise ValueError(
                'Lambda function execution resulted in error',
                {"ResponseMetadata": response.get("ResponseMetadata"), "Payload": payload},
            )
        self.log.info('Lambda function invocation succeeded: %r', response.get("ResponseMetadata"))
        return payload
```

ferruzzi commented on docs/apache-airflow-providers-amazon/operators/lambda.rst:
```diff
@@ -0,0 +1,54 @@
+
+.. _howto/operator:AwsLambdaInvokeFunctionOperator:
+
+AWS Lambda Invoke Function Operator
```
Nitpick. I'm working on standardizing these doc files a bit, if you don't mind.

```suggestion
AWS Lambda Operators
```

schirag1993 replied:
Not at all. If you have a draft of the docs standardisation (I can view), happy to add anything else I can find.

ferruzzi commented on airflow/providers/amazon/aws/operators/aws_lambda.py:
```diff
@@ -0,0 +1,99 @@
+        invocation_type: Optional[str] = None,
+        client_context: Optional[str] = None,
+        payload: Optional[str] = None,
+        aws_conn_id='aws_default',
```
```suggestion
        aws_conn_id: str = 'aws_default',
```

schirag1993 committed:
update documentaiton code based on feedback

schirag1993 committed:
add AWSLambdaInvokeFunction operator example DAG
